### PR TITLE
Add singleton

### PR DESCRIPTION
**Pull request by Heavy32**

### CODE REVIEW

File: Singleton/AccountCounter.cs [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
+        }
 
-        private AccountCounter() { }
+        public void WriteInLog(string logMessage)
+        {
+            DateTime now = DateTime.Now;
+            logger.Write("Log ID: " + logID + " Date:"  + now + " " + logMessage);
+        }
+
+        public void FinishLogger()
```
Comment (alexzaitzev) [inline Singleton/AccountCounter.cs]: Itæs better to implement Disposable pattern here: https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

File: Singleton/AccountCounter.cs [hunk 2]
```diff
@@ -1,16 +1,32 @@
...
-        private static readonly Lazy<AccountCounter> lazy = new Lazy<AccountCounter>(() => new AccountCounter());
+        private static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private int logID = 1;
+        private StreamWriter logger;
 
-        public static AccountCounter Instance { get { return lazy.Value; } }
-        public int count = 0;
+        private LoggerSingleton()
+        {
+            logger = new StreamWriter("logger.txt");
```
Comment (alexzaitzev) [inline Singleton/AccountCounter.cs]: Did you try to execute this code?
```csharp
LoggerSingleton.Instance.WriteInLog("a");
LoggerSingleton.Instance.WriteInLog("b");
```

File: Singleton/AccountCounter.cs [hunk 3]
```diff
@@ -1,16 +1,32 @@
...
 
 namespace Singleton
 {
-    public sealed class AccountCounter
+    public sealed class LoggerSingleton
     {
-        private static readonly Lazy<AccountCounter> lazy = new Lazy<AccountCounter>(() => new AccountCounter());
+        private static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private int logID = 1;
+        private StreamWriter logger;
```
Comment (alexzaitzev) [inline Singleton/AccountCounter.cs]: Field should be readonly

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Please have a look on my comments

File: Singleton/Account.cs [hunk 4]
```diff
@@ -5,14 +5,14 @@ namespace Singleton
 {
     class Account
     {
-        public string Name { get; private set; }
-        public int Age { get; private set; }
+        //public string Name { get; private set; }
```
Comment (alexzaitzev) [inline Singleton/Account.cs]: If you don't need it - delete

File: Singleton/LoggerSingleton.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Singleton
+{
+    public sealed class LoggerSingleton : IDisposable
+    {
+        private static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private int logID = 1;
+        private StreamWriter logger;
+        private bool disposed = false;
+        private readonly SafeHandle handle = new SafeFileHandle(IntPtr.Zero, true);
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Why it's needed?

File: Singleton/LoggerSingleton.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                Logger.Dispose();
+            }
+
+            disposed = true;
+        }
+
+        public static LoggerSingleton Instance { get => instance.Value; }
+        public int LogID { get => logID; set => logID = value; }
+        public StreamWriter Logger { get => logger; set => logger = value; }
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Why did you exposed internal logger?

File: Singleton/LoggerSingleton.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+using System.IO;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Singleton
+{
+    public sealed class LoggerSingleton : IDisposable
+    {
+        private static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private int logID = 1;
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Can be converted to auto property

File: Singleton/LoggerSingleton.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        private StreamWriter logger;
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Can be converted to auto propert as well

File: Singleton/LoggerSingleton.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        private bool disposed = false;
+        private readonly SafeHandle handle = new SafeFileHandle(IntPtr.Zero, true);
+
+        private LoggerSingleton() { }
+
+        public void StartLogger()
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Example: you have 2 threads where you need to log. You need to run `StartLogger` in both -> The same probelm as before

File: Singleton/LoggerSingleton.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            Logger = new StreamWriter("Logger1.txt");
+        }
+
+        public void WriteInLog(string logMessage)
+        {
+            DateTime now = DateTime.Now;
+            Logger.Write("Log ID: " + LogID + " Date:"  + now + " " + logMessage + "\n");
+        }
+
+        public void FinishLogger()
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: To better understand Singleton pattern write an example where you write log in two threads

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Address the comments needed

File: Singleton/Program.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+using System.Text;
+using System.Threading;
+
+namespace Singleton
+{
+    class Program
+    {
+        static void Main()
+        {
+            (new Thread(() =>
```
Comment (alexzaitzev) [inline Singleton/Program.cs]: `Task.Run(...)` instead of `new Thread()`

File: Singleton/LoggerSingleton.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            if (disposing)
+            {
+                Logger.Dispose();
+            }
+
+            disposed = true;
+        }
+
+        public int LogID { get; set; } = 1;
+        public StreamWriter Logger { get; set; }
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Why is it still `public` ?  :)

File: Singleton/Program.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+            {
+                LoggerSingleton.instance.Value.WriteInLog("Sup");
+            })).Start();
+
+            LoggerSingleton.instance.Value.WriteInLog("Nigga");
+            Console.Read();
+            LoggerSingleton.instance.Value.FinishLogger();
```
Comment (alexzaitzev) [inline Singleton/Program.cs]: It's kind of cheating, let's do without `Console.Read` and `FinishLogger`

File: Singleton/AccountCounter.cs (hunk 1 above)
Comment (Heavy32) [reply Singleton/AccountCounter.cs]: I've used "using" section for Logger logic, is it the same?

File: Singleton/AccountCounter.cs (hunk 2 above)
Comment (Heavy32) [reply Singleton/AccountCounter.cs]: done

File: Singleton/AccountCounter.cs (hunk 3 above)
Comment (Heavy32) [reply Singleton/AccountCounter.cs]: done

File: Singleton/Account.cs (hunk 4 above)
Comment (Heavy32) [reply Singleton/Account.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 5 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: deleted

File: Singleton/LoggerSingleton.cs (hunk 6 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: I've changed the logic

File: Singleton/LoggerSingleton.cs (hunk 7 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 8 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 9 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: Problem's been solved

File: Singleton/LoggerSingleton.cs (hunk 10 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/Program.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+        static void Main()
+        {
+            Task t1 = new Task(() => LoggerSingleton.instance.Value.WriteInLog("Hello"));
+            Task t2 = new Task(() => LoggerSingleton.instance.Value.WriteInLog("Leha"));
+
+            t1.Start();
+            t1.Wait();
+
+            t2.Start();
+            t2.Wait();
```
Comment (alexzaitzev) [inline Singleton/Program.cs]: You are cheating again. As far as I remeber test code should look like
```
Task.Run(() =>
{
	LoggerSingleton.instance.Value.WriteInLog("Hello");
});

LoggerSingleton.instance.Value.WriteInLog("Leha");
```

File: Singleton/LoggerSingleton.cs [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+        public void WriteInLog(string logMessage)
+        {
+            DateTime now = DateTime.Now;
+
+            string writePath = @"~\Singleton\Logger.txt";
+
+            using StreamWriter logger = new StreamWriter(writePath, true, System.Text.Encoding.Default);
+                logger.WriteLine("Log ID: " + LogID++ + " Date:" + now + " " + logMessage);
+
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Extra empty line

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Changes needed

File: Singleton/Program.cs (hunk 11 above)
Comment (Heavy32) [reply Singleton/Program.cs]: Done

File: Singleton/LoggerSingleton.cs (hunk 12 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/Program.cs (hunk 13 above)
Comment (Heavy32) [reply Singleton/Program.cs]: Done

File: Singleton/Program.cs (hunk 14 above)
Comment (Heavy32) [reply Singleton/Program.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 15 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/Logger.txt [hunk 16]
```diff
@@ -0,0 +1,2 @@
+Log ID: 1 Date:13.05.2020 23:34:45 Leha
```
Comment (alexzaitzev) [inline Singleton/Logger.txt]: Do we need this file?

File: Singleton/LoggerSingleton.cs [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    public sealed class LoggerSingleton
+    {
+        public static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private static readonly object syncRoot = new object();
+
+        private LoggerSingleton() { }
+
+        public void WriteInLog(string logMessage)
+        {
+            DateTime now = DateTime.Now;
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: Can we inline this?

File: Singleton/LoggerSingleton.cs [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+            string writePath = @"C:\Users\eBasher\Documents\GitHub\Patterns\Singleton\Logger.txt";
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: We need to init it every time? It can be made as a field

File: Singleton/LoggerSingleton.cs [hunk 19]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.IO;
+
+namespace Singleton
+{
+    public sealed class LoggerSingleton
+    {
+        public static readonly Lazy<LoggerSingleton> instance = new Lazy<LoggerSingleton>(() => new LoggerSingleton());
+        private static readonly object syncRoot = new object();
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: `static` is not needed I guess

File: Singleton/LoggerSingleton.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+            string writePath = @"C:\Users\eBasher\Documents\GitHub\Patterns\Singleton\Logger.txt";
+
+            lock (syncRoot)
+            {
+                using StreamWriter logger = new StreamWriter(writePath, true, System.Text.Encoding.Default);
+                    logger.WriteLine("Log ID: " + LogID++ + " Date:" + now + " " + logMessage);
+            }           
+        }
+
+        public int LogID { get; set; } = 1;
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: What if it will be more than 2 147 483 647 records?

File: Singleton/Program.cs [hunk 21]
```diff
@@ -0,0 +1,22 @@
+﻿using System.IO;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System;
+
+namespace Singleton
+{
+    class Program
```
Comment (alexzaitzev) [inline Singleton/Program.cs:5]: I ran it and got `DirectoryNotFoundException`.

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Comments need to be addressed

File: Singleton/Program.cs (hunk 21 above)
Comment (Heavy32) [reply Singleton/Program.cs:5]: Now the logger.txt file is saved in "bin" folder.
Not sure If it's better, but it has to work at least

File: Singleton/LoggerSingleton.cs (hunk 20 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: Changed int to long

File: Singleton/LoggerSingleton.cs (hunk 18 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 17 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/LoggerSingleton.cs (hunk 19 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: done

File: Singleton/Logger.txt (hunk 16 above)
Comment (Heavy32) [reply Singleton/Logger.txt]: I left it to let you see the way the program works with a file created earlier, but okay, I've removed

File: Singleton/LoggerSingleton.cs [hunk 22]
```diff
@@ -6,23 +6,18 @@ namespace Singleton
...
         {
-            DateTime now = DateTime.Now;
-
-            string writePath = @"C:\Users\eBasher\Documents\GitHub\Patterns\Singleton\Logger.txt";
-
             lock (syncRoot)
             {
-                using StreamWriter logger = new StreamWriter(writePath, true, System.Text.Encoding.Default);
-                    logger.WriteLine("Log ID: " + LogID++ + " Date:" + now + " " + logMessage);
+                using StreamWriter logger = new StreamWriter(@"Logger.txt", true, System.Text.Encoding.Default);
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: @ needed?

File: Singleton/LoggerSingleton.cs [hunk 23]
```diff
@@ -6,23 +6,18 @@ namespace Singleton
...
+                    logger.WriteLine("Log ID: " + LogID++ + " Date:" + DateTime.Now + " " + logMessage);
             }           
         }
 
-        public int LogID { get; set; } = 1;
+        public long LogID { get; set; } = 1;
```
Comment (alexzaitzev) [inline Singleton/LoggerSingleton.cs]: What if `long.MaxValue + 1`? :)

File: Singleton/Singleton.csproj [hunk 24]
```diff
@@ -1,8 +1,12 @@
...
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp3.1</TargetFramework>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <OutputPath></OutputPath>
+  </PropertyGroup>
```
Comment (alexzaitzev) [inline Singleton/Singleton.csproj]: Do we need it?

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Small comments

File: Singleton/LoggerSingleton.cs (hunk 23 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: Done

File: Singleton/LoggerSingleton.cs (hunk 22 above)
Comment (Heavy32) [reply Singleton/LoggerSingleton.cs]: Done

File: Singleton/Singleton.csproj (hunk 24 above)
Comment (Heavy32) [reply Singleton/Singleton.csproj]: done

**Review by alexzaitzev — APPROVED** [review]

**Comment by alexzaitzev** [comment]
Looks like you have conflicts :)

**Pull request merged**